controllers/depresolver/depresolver_config.go

```diff
@@ -55,9 +57,26 @@ const (
 	MetricsAddressKey              = "METRICS_ADDRESS"
 )
 
+type OldVar = string
+type NewVar struct {
+	Name string
+	Msg  string
+}
+
+var deprecated = map[OldVar]NewVar{
+	"EDGE_DNS_SERVER": NewVar{
+		Name: "EDGE_DNS_SERVERS",
+		Msg:  "Pass the hostname or IP address as comma-separated list",
+	},
+	"EDGE_DNS_PORT": NewVar{
+		Name: "EDGE_DNS_SERVERS",
+		Msg:  "Port is an optional in the comma-separated list of dns edge servers, in following form: dns1:53,dns2",
+	},
+}
+
```

Comment:
> OldVar is practically n ot used, map[string]NewVar

it's not, but it's more readable this way, imho... if anyone sees the type it's quite obvious that it's a mapping from old variable name to the new variable name

> Also NewVar, OldVar are not exposed, so can be private

good point

> EDGE_DNS_SERVER,EDGE_DNS_PORT, should be constants as it is defined in the rest of the package & EDGE_DNS_SERVER is used twice

whatever, I am putting them back, I wanted to prepare the stuff for their removal in the future. That's how deprecation logic should work in general, if you think it needs to be there forever, I can add them back and group them together under some deprecated comment.. or under their own const block

> but can't we just define public function retreiving deprecated stuffs

nah, this approach would print the message even though the deprecated env var was not being used. There must be some data (map, slice whatever) saying this particular env var is deprecated and we need to check if it's set by the user. I guess, I can move it to the `findDeprecations` function itself, but when called multiple times it will be doing the same stuff over and over again (not big problem in our context, because it's called only once during the startup), like this:

```golang
func (dr *DependencyResolver) findDeprecations(config *Config) (deprecations []string) {
	type oldVar = string
	type newVar struct {
		Name string
		Msg  string
	}
	var deprecated = map[oldVar]newVar{
		"EDGE_DNS_SERVER": newVar{
			Name: "EDGE_DNS_SERVERS",
			Msg:  "Pass the hostname or IP address as comma-separated list",
		},
		"EDGE_DNS_PORT": newVar{
			Name: "EDGE_DNS_SERVERS",
			Msg:  "Port is an optional in the comma-separated list of dns edge servers, in following form: dns1:53,dns2",
		},
	}

	for k, v := range deprecated {
		if os.Getenv(k) != "" {
			deprecations = append(deprecations, fmt.Sprintf("'%s' has been deprecated, use %s instead. Details: %s", k, v.Name, v.Msg))

			// remove the following code once the EDGE_DNS_SERVER and EDGE_DNS_PORT are not supported
			if k == "EDGE_DNS_SERVER" {
				// old contract, let's transform it to the new api -> will be also validated
				edgeDNSHost := os.Getenv(k)
				edgeDNSPort, _ := env.GetEnvAsIntOrFallback("EDGE_DNS_PORT", 53)
				config.EdgeDNSServers = []utils.DNSServer{{Host: edgeDNSHost, Port: edgeDNSPort}}
			}
		}
	}
	return
}
```

I don't like this ^ It's long and I don't like mixing static data w/ behavior so I'd leave it under the `consts` because it's more readable this way, imho. Having a function returning the map would also work and calling it from `findDeprecations` but what would be the point if it's returning the same data over and over again.

Reply:
Ok, I understand your reasoning, `EDGE_DNS_SERVER` is a dynamic variable and the deprecation list should be returned depending on how the individual variables are deprecated in the future.

`EDGE_DNS_SERVER`, `EDGE_DNS_SERVER_PORT` can be used as a before, and there is no need to delete them (to make thing deprecated, ).  This implies several things:
 - we won't have to deal with breaking change after all, at least not at the moment
 - we have added a new variable EDGE_DNS_SERVERS ; if EDGE_DNS_SERVER, EDGE_DNS_SERVER_PORT are filled in, we just extend EDGE_DNS_SERVERS with another value 
 - `findDeprecations` can be simplified a lot, like this:
```go
func (dr *DependencyResolver) findDeprecations(config *Config) (deprecations []string) {
	if dr.config.EdgeDNSServer != "" {
		deprecations = append(deprecations, fmt.Sprintf("%s has been deprecated, use %s instead .... ", EdgeDNSServerKey, EdgeDNSServerKey)
		deprecations = append(deprecations, fmt.Sprintf("%s has been deprecated, use %s instead ... ", EdgeDNSServerPortKey, EdgeDNSServerKey)
		}
	}
       // if something else is deprecated, deprecations list will be altered here...
return
}
```
 
`EDGE_DNS_SERVER, EDGE_DNS_SERVER_PORT, EDGE_DNS_SERVERS `- now we have three live envvars, we don't have breaking changes, maybe we need to change validation and tests a bit.


Usually, when something is deprecate din golang, we use `Deprecated comments`, see: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go#L65-L71